forte/processors/data_augment/algorithms/embedding_similarity_augmenter.py

```diff
@@ -0,0 +1,104 @@
+# Copyright 2020 The Forte Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import random
+from typing import Dict, List
+import numpy as np
+from tensorflow import gfile
+from texar.tf.data import Embedding, load_glove
+from forte.processors.data_augment.algorithms.base_augmenter import BaseDataAugmenter
+
+__all__ = [
+    "EmbeddingSimilarityAugmenter",
+]
+
+random.seed(0)
+
+emb_types = ['glove']
+
+
+def load_glove_vocab(filename):
+    vocab = {}
+    idx_to_word = {}
+    with gfile.GFile(filename) as fin:
+        for line in fin:
+            vec = line.strip().split()
+            if len(vec) == 0:
+                continue
+            word = vec[0]
+            word_idx = len(vocab)
+            vocab[word] = word_idx
+            idx_to_word[word_idx] = word
+    return vocab, idx_to_word
+
+def l2_norm(word_vecs):
+    norm = np.sqrt((word_vecs * word_vecs).sum(axis=1))
+    return word_vecs / norm[:, np.newaxis]
+
+class EmbeddingSimilarityAugmenter(BaseDataAugmenter):
+    r"""
+    This class is a data augmenter leveraging pre-trained word
+    embeddings, such as word2vec and glove, to replace the input
+    word with another word with similar word embedding.
+    By default, the replacement word is randomly chosen from the
+    top k words with the most similar embeddings.
+
+    Args:
+        emb_path: Path to the pretrained embedding file
+        emb_type: Type of embedding. E.g. glove
+        emb_dim: Dimension of the embedding
+    """
+    def __init__(self, emb_path, emb_type, emb_dim, configs: Dict[str, str]):
```

Comment:
1. We should be consistent in terms of the configuration, you can put everything in the configs dict, or you can put all configs as named parameters. The latter method should be better because you can pass a dict as named parameters.
2. We will implement the embedded using Texar, hence we should pass in the embedder here directly.

Reply:
2. Okay, I see. In this class, we assume the Embedder/Embedding is already initialized. 

So do we separate the "reading pre-trained word embedding" part out of this class, and give the user some util functions (e.g. whose input is a Glove format text file) to do that themselves?

Since our input is text, we also need a word-to-index mapping and an index-to-word mapping, right?

Reply:
Yes, we can directly pass in an `embedder` object, which the user can freely choose. We shouldn't implement again though,  
 for word embedding, we should use the Texar's Embedding class (https://github.com/asyml/texar-pytorch/blob/master/texar/torch/data/embedding.py#L103). In the future we may need to use contextual embeddings such as Bert, we should use the Texar Embedder class (https://github.com/asyml/texar-pytorch/tree/master/texar/torch/modules/pretrained).

Reply:
For the word-to-index and index-to-word mapping, yes we should have something for that, that is called the Vocabulary. So unfortunately I find that the Vocab implementation in texar-pytorch is not very complete, it can only load a vocabulary file. https://github.com/asyml/texar-pytorch/blob/master/texar/torch/data/vocabulary.py

However, I don't want to create too many copies of a simple `Vocabulary` implementation. What do you think here?

Reply:
Oh okay. Maybe we could make another util function to read in Glove or Word2vec file and generate a vocabulary file. Would that be a good solution?